File: client/android/src/main/java/io/codemagic/patch/CodemagicPatchState.kt
package io.codemagic.patch

import org.json.JSONObject

internal data class CodemagicPatchPackagePointer(val packageHash: String) {
  fun toJson(): JSONObject =
      JSONObject().put("package_hash", packageHash)
}

internal data class CodemagicPatchFailedInstall(
    val packageHash: String,
    val reason: String,
    val failedAt: String
) {
  fun toJson(): JSONObject =
      JSONObject()
        .put("package_hash", packageHash)
        .put("reason", reason)
        .put("failed_at", failedAt)
}

internal data class CodemagicPatchState(
    var current: CodemagicPatchPackagePointer? = null,
    var previous: CodemagicPatchPackagePointer? = null,
    var pending: CodemagicPatchPackagePointer? = null,
    var failedInstall: CodemagicPatchFailedInstall? = null,
    var pendingStarted: String? = null
) {
  val packageHashes: List<String>
    get() = listOfNotNull(
      pending?.packageHash,
      current?.packageHash,
      previous?.packageHash
    )

  fun toJson(): JSONObject {
    val json = JSONObject()
    current?.let { json.put("current", it.toJson()) }
    previous?.let { json.put("previous", it.toJson()) }
    pending?.let { json.put("pending", it.toJson()) }
    failedInstall?.let { json.put("failed_install", it.toJson()) }
    pendingStarted?.let { json.put("pending_started", it) }
    return json
  }

  companion object {
    fun fromJson(json: JSONObject, isSafePackageHash: (String) -> Boolean): CodemagicPatchState {
      fun pointer(name: String): CodemagicPatchPackagePointer? {
        val hash = json.optJSONObject(name)?.optString("package_hash") ?: return null
        return if (isSafePackageHash(hash)) CodemagicPatchPackagePointer(hash) else null
      }

      val failed = json.optJSONObject("failed_install")?.let {
        val hash = it.optString("package_hash")
        if (isSafePackageHash(hash)) {
          CodemagicPatchFailedInstall(
            packageHash = hash,
            reason = it.optString("reason"),
            failedAt = it.optString("failed_at")
          )
        } else {
          null
        }
      }
      val pendingStarted = json.optString("pending_started")
        .takeIf { isSafePackageHash(it) }

      return CodemagicPatchState(
        current = pointer("current"),
        previous = pointer("previous"),
        pending = pointer("pending"),
        failedInstall = failed,
        pendingStarted = pendingStarted
      )
    }
  }
}

File: client/ios/CodemagicPatchState.swift
import Foundation

struct CodemagicPatchPackagePointer: Codable {
  var packageHash: String

  enum CodingKeys: String, CodingKey {
    case packageHash = "package_hash"
  }
}

struct CodemagicPatchFailedInstall: Codable {
  var packageHash: String
  var reason: String
  var failedAt: String

  enum CodingKeys: String, CodingKey {
    case packageHash = "package_hash"
    case reason
    case failedAt = "failed_at"
  }
}

struct CodemagicPatchState: Codable {
  var current: CodemagicPatchPackagePointer? = nil
  var previous: CodemagicPatchPackagePointer? = nil
  var pending: CodemagicPatchPackagePointer? = nil
  var failedInstall: CodemagicPatchFailedInstall? = nil
  var pendingStarted: String? = nil
  // Optional so state written by an SDK older than the launch-attempt budget
  // still decodes: the synthesized decoder would throw on a missing key for a
  // non-optional, and a throw here resets every lifecycle pointer.
  var pendingStartCount: Int? = nil

  enum CodingKeys: String, CodingKey {
    case current
    case previous
    case pending
    case failedInstall = "failed_install"
    case pendingStarted = "pending_started"
    case pendingStartCount = "pending_start_count"
  }

  var packageHashes: [String] {
    [pending, current, previous].compactMap { $0?.packageHash }
  }

  /// Launches already charged against the pending package's attempt budget.
  var pendingLaunchAttempts: Int {
    max(pendingStartCount ?? 0, 0)
  }

  /**
   Charges one launch against `hash`'s attempt budget. A different pending hash
   restarts the budget, so a freshly installed package always gets the full
   allowance.
   */
  mutating func chargePendingLaunch(_ hash: String) {
    pendingStartCount = pendingStarted == hash ? pendingLaunchAttempts + 1 : 1
    pendingStarted = hash
  }

  /**
   Clears the launch-attempt tracking pair. The marker and its counter are only
   ever meaningful together, so they are always cleared together.
   */
  mutating func clearPendingLaunchTracking() {
    pendingStarted = nil
    pendingStartCount = nil
  }

  func sanitized() -> CodemagicPatchState {
    let safePendingStarted = pendingStarted?.takeIfSafePackageHash()
    return CodemagicPatchState(
      current: current.sanitized(),
      previous: previous.sanitized(),
      pending: pending.sanitized(),
      failedInstall: failedInstall.sanitized(),
      pendingStarted: safePendingStarted,
      pendingStartCount: safePendingStarted == nil ? nil : pendingStartCount
    )
  }
}

private extension Optional where Wrapped == CodemagicPatchPackagePointer {
  func sanitized() -> CodemagicPatchPackagePointer? {
    guard let pointer = self,
          CodemagicPatchStorage.isSafePackageHash(pointer.packageHash) else {
      return nil
    }
    return pointer
  }
}

private extension Optional where Wrapped == CodemagicPatchFailedInstall {
  func sanitized() -> CodemagicPatchFailedInstall? {
    guard let failed = self,
          CodemagicPatchStorage.isSafePackageHash(failed.packageHash) else {
      return nil
    }
    return failed
  }
}

private extension String {
  func takeIfSafePackageHash() -> String? {
    CodemagicPatchStorage.isSafePackageHash(self) ? self : nil
  }
}
